File: make_dummy_inputs.js
/*
 * make_dummy_inputs.js — Node script writing synthetic input_dummy/*.xlsx:
 * DataMart + one zone-data file per zone (MAZ, NAZ, SAZ, AFR, EUR, APAC, GHQ) +
 * Saviynt / O365 / Aurora / BSC / CEO. Labeled emails drive every pipeline filter.
 */
const fs = require("fs");
const path = require("path");
const io = require("./web/pipeline/io.js");

const ZONES = ["MAZ", "NAZ", "SAZ", "AFR", "EUR", "APAC", "GHQ"];

const REQUIRED_COLUMNS = [
  "Zone", "Country", "Global Employee ID", "Local Employee ID", "Employee Name",
  "Employee Status", "Worker Type", "Employee Group", "Management Level",
  "First Hire Date", "Last Hire Date", "Position Name", "Job Family Group",
  "Job Family", "Job Profile Description", "ABI Entity 2",
  "Macro Entity Level 2 (Zone)", "text before Email", "Employee Email"
];

function mk(over) {
  const r = {};
  REQUIRED_COLUMNS.forEach(function (c) { r[c] = ""; });
  return Object.assign(r, over);
}

function datamartTable() {
  const rows = [
    mk({ "Employee Email": "alice@corp.com", Zone: "NAZ", "Employee Name": "Alice A",
      "Global Employee ID": "G1", "Local Employee ID": "G1", "Employee Status": "Active" }),
    mk({ "Employee Email": "alice@corp.com", Zone: "NAZ", "Employee Name": "Alice A",
      "Global Employee ID": "G1", "Local Employee ID": "G1", "Employee Status": "Active" }),
    mk({ "Employee Email": "", Zone: "NAZ", "Employee Name": "Blank B",
      "Global Employee ID": "G2", "Local Employee ID": "G2" }),
    mk({ "Employee Email": "noemail99", Zone: "NAZ", "Employee Name": "NoEmail N",
      "Global Employee ID": "G3", "Local Employee ID": "G3" }),
    mk({ "Employee Email": "maz_ok@corp.com", Zone: "MAZ", "Employee Name": "Maz Ok",
      "Global Employee ID": "G4", "Local Employee ID": "G4" }),
    mk({ "Employee Email": "maz_bad@corp.com", Zone: "MAZ", "Employee Name": "Maz Bad",
      "Global Employee ID": "G5", "Local Employee ID": "G5" }),
    mk({ "Employee Email": "ot@corp.com", Zone: "NAZ", "Employee Name": "OT User",
      "Global Employee ID": "G6", "Local Employee ID": "G6",
      "Job Family Group": "SUPPLY", "Job Family": "Plant Management",
      "Job Profile Description": "Brewery Plant Manager" }),
    mk({ "Employee Email": "ceo_target@corp.com", Zone: "NAZ", "Employee Name": "Ceo Target",
      "Global Employee ID": "G7", "Local Employee ID": "G7" })
  ];
  return { columns: REQUIRED_COLUMNS.slice(), rows: rows };
}

// Action=OK rows per zone. NAZ grants OK to alice/ot/ceo_target; MAZ to maz_ok.
const ZONE_OK = {
  MAZ: [
    { id: "G4", name: "Maz Ok", action: "OK" },
    { id: "G5", name: "Maz Bad", action: "No" }
  ],
  NAZ: [
    { id: "G1", name: "Alice A", action: "OK" },
    { id: "G6", name: "OT User", action: "OK" },
    { id: "G7", name: "Ceo Target", action: "OK" }
  ]
};

function zoneValidationTable(zone) {
  const rows = (ZONE_OK[zone] || []).map(function (r) {
    return { "Global Employee ID": r.id, "Local Employee ID": r.id,
      "Employee Name": r.name, "Action": r.action };
  });
  return io.tableFromObjects(
    ["Global Employee ID", "Local Employee ID", "Employee Name", "Action"], rows);
}

function buildDummyInputs(outDir) {
  fs.mkdirSync(outDir, { recursive: true });
  const paths = { datamart: path.join(outDir, "Data Mart for userbase creation.xlsx") };
  const write = function (p, bytes) { fs.writeFileSync(p, Buffer.from(bytes)); };

  write(paths.datamart, io.tableToXlsxBytes(datamartTable(), "DataMart"));

  // one file per zone; MAZ also carries an "add to the list" sheet
  ZONES.forEach(function (zone) {
    const key = "zone_" + zone;
    paths[key] = path.join(outDir, zone + " zone data.xlsx");
    const sheets = [{ name: zone, table: zoneValidationTable(zone) }];
    if (zone === "MAZ") {
      sheets.push({ name: "add to the list", table: io.tableFromObjects(
        ["Zone", "Country", "Global Employee ID", "Local Employee ID", "Employee Name",
          "Employee Status", "Management Level", "First Hire Date", "Last Hire Date",
          "Position Name", "Employee Email"],
        [{ Zone: "MAZ", Country: "X", "Global Employee ID": "G8", "Local Employee ID": "G8",
          "Employee Name": "Maz Add", "Employee Status": "Active",
          "Employee Email": "maz_add@corp.com" }]) });
    }
    write(paths[key], io.tablesToXlsxBytes(sheets));
  });

  paths.saviynt = path.join(outDir, "Saviynt.xlsx");
  write(paths.saviynt, io.tableToXlsxBytes(io.tableFromObjects(
    ["User Email", "SSO UPN"], [{ "User Email": "alice@corp.com", "SSO UPN": "alice_upn" }]), "Saviynt"));

  paths.o365 = path.join(outDir, "O365.xlsx");
  write(paths.o365, io.tablesToXlsxBytes([{ name: "Export", table: io.tableFromObjects(
    ["Mail", "UserPrincipalName"], [{ Mail: "alice@corp.com", UserPrincipalName: "alice@ad" }]) }]));

  paths.aurora = path.join(outDir, "Aurora Users.xlsx");
  write(paths.aurora, io.tablesToXlsxBytes([{ name: "Aurora Userbase", table: io.tableFromObjects(
    ["NAME", "E-MAIL", "reverse lookup"], [
      { NAME: "Alice A", "E-MAIL": "alice@corp.com", "reverse lookup": "found" },
      { NAME: "Aurora New", "E-MAIL": "aurora_new@corp.com", "reverse lookup": "not found" }
    ]) }]));

  paths.bsc = path.join(outDir, "BSC users.xlsx");
  write(paths.bsc, io.tablesToXlsxBytes([{ name: "main", table: io.tableFromObjects(
    ["Email - Primary Work", "Zone", "reverse lookup"], [
      { "Email - Primary Work": "alice@corp.com", Zone: "NAZ", "reverse lookup": "found" },
      { "Email - Primary Work": "bsc_new@corp.com", Zone: "MAZ", "reverse lookup": "not found" }
    ]) }]));

  paths.ceo = path.join(outDir, "CEO Minus 1 and 2 - 2026 1.xlsx");
  write(paths.ceo, io.tablesToXlsxBytes([
    { name: "19 Feb 2026", table: io.tableFromObjects(["Manager", "Mail ID"], [{ Manager: "Old", "Mail ID": "old@corp.com" }]) },
    { name: "25 May 2026", table: io.tableFromObjects(["Manager", "Mail ID"], [{ Manager: "Boss", "Mail ID": "ceo_target@corp.com" }]) }
  ]));

  return paths;
}

module.exports = { buildDummyInputs, REQUIRED_COLUMNS, ZONES };

if (require.main === module) {
  const out = path.join(__dirname, "input_dummy");
  const paths = buildDummyInputs(out);
  console.log("Wrote dummy inputs:");
  Object.keys(paths).forEach(function (k) { console.log(" - " + k + ": " + paths[k]); });
}

File: scripts/make_dummy_inputs.py
"""Deterministic dummy input generator for manual UI testing and the e2e test."""
import sys
from pathlib import Path

import pandas as pd

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from engine import config  # noqa: E402


def _email(i):
    if i <= 5:
        return ""
    if i <= 8:
        return f"noemail.user{i:03d}@abi.com"
    if i <= 10:
        return f"user{i:03d}.abi.com"  # missing @
    if i <= 14:
        return "dup@abi.com"
    return f"user{i:03d}@abi.com"


def _zone(i):
    return {0: "MAZ", 1: "SAZ"}.get(i % 3, "Other")


def build(out_dir) -> dict:
    out_dir = Path(out_dir)
    out_dir.mkdir(parents=True, exist_ok=True)
    n = 60
    rows = []
    for i in range(1, n + 1):
        r = {c: f"{c[:3]}{i}" for c in config.REQUIRED_COLUMNS}
        r["Zone"] = _zone(i)
        r["Employee Email"] = _email(i)
        r["Employee Name"] = f"User {i:03d}"
        if i in (15, 16, 17):
            r["Job Family Group"] = "SUPPLY"
            r["Job Family"] = "Plant Management"
            r["Job Profile Description"] = "Brewery Plant Manager"
        else:
            r["Job Family Group"] = "SALES"
            r["Job Family"] = "Commercial"
            r["Job Profile Description"] = "Sales Rep"
        rows.append(r)
    dm = pd.DataFrame(rows)
    dm["Extra Junk Column"] = "junk"
    files = {}

    p = out_dir / "datamart.xlsx"
    dm.to_excel(p, index=False)
    files["datamart"] = p

    valid = dm[(dm["Employee Email"].str.contains("@"))
               & (~dm["Employee Email"].str.contains("noemail"))]

    maz_emails = valid[valid["Zone"] == "MAZ"]["Employee Email"].drop_duplicates().tolist()
    maz_actions = ["Terminate" if e == maz_emails[-1] else "OK" for e in maz_emails]
    p = out_dir / "zone_MAZ.xlsx"
    with pd.ExcelWriter(p) as w:
        pd.DataFrame({"Employee Email": maz_emails, "Action": maz_actions}).to_excel(
            w, sheet_name="MAZ", index=False)
        pd.DataFrame({
            "Employee Email": ["maz.add1@abi.com", "maz.add2@abi.com"],
            "Employee Name": ["MAZ Add One", "MAZ Add Two"],
            "Zone": ["MAZ", "MAZ"],
        }).to_excel(w, sheet_name="add to the list", index=False)
    files["zone_MAZ"] = p

    saz_emails = valid[valid["Zone"] == "SAZ"]["Employee Email"].drop_duplicates().tolist()
    p = out_dir / "zone_SAZ.xlsx"
    pd.DataFrame({"Employee Email": saz_emails, "Action": ["OK"] * len(saz_emails)}).to_excel(
        p, sheet_name="SAZ", index=False)
    files["zone_SAZ"] = p

    o365_rows = [(e, f"upn{i}@ad.abi.com")
                 for i, e in enumerate(valid["Employee Email"].drop_duplicates(), 1)
                 if i % 2 == 0]
    p = out_dir / "o365.xlsx"
    pd.DataFrame(o365_rows, columns=["Mail", "UserPrincipalName"]).to_excel(p, index=False)
    files["o365"] = p

    sav_rows = [(e, f"sso{i}@abi.com")
                for i, e in enumerate(valid["Employee Email"].drop_duplicates(), 1)
                if i % 3 == 0]
    p = out_dir / "saviynt.xlsx"
    pd.DataFrame(sav_rows, columns=["User Email", "SSO UPN"]).to_excel(p, index=False)
    files["saviynt"] = p

    some = valid["Employee Email"].drop_duplicates().tolist()[:3]
    p = out_dir / "aurora.xlsx"
    pd.DataFrame({
        "E-MAIL": some + ["aurora.new1@abi.com", "aurora.new2@abi.com"],
        "NAME": [f"N{i}" for i in range(len(some))] + ["Aurora New1", "Aurora New2"],
    }).to_excel(p, sheet_name="Aurora Userbase", index=False)
    files["aurora"] = p

    p = out_dir / "bsc.xlsx"
    pd.DataFrame({"Email - Primary Work": some[:2] + ["bsc.new1@abi.com"]}).to_excel(
        p, sheet_name="main", index=False)
    files["bsc"] = p

    p = out_dir / "ceo.xlsx"
    with pd.ExcelWriter(p) as w:
        pd.DataFrame({"Mail ID": ["stale@abi.com"]}).to_excel(
            w, sheet_name="19 Feb 2026", index=False)
        pd.DataFrame({"Mail ID": ["user018@abi.com", "upn2@ad.abi.com"]}).to_excel(
            w, sheet_name="25 May 2026", index=False)
    files["ceo"] = p

    return files


if __name__ == "__main__":
    out = build(Path(__file__).resolve().parent.parent / "input_dummy")
    for slot, path in out.items():
        print(f"{slot}: {path}")
